### PR TITLE
Support ARM architecture

### PR DESCRIPTION
**Pull request by lecopzer**
I'm boring in this Chinese New Year, so I try to write some code for fun.

Now min-dl works in cross compile for arm in following cross compiler 
```
make CROSS_COMPILE=arm-linux-gnueabi- check
 > gcc version 5.4.0 20160609 (Ubuntu/Linaro 5.4.0-6ubuntu1~16.04.4)
qemu-arm  -L /usr/arm-linux-gnueabi/ loader
 > qemu-arm version 2.5.0
```
I think we can also move all the system-specific assembly macro to independent header file
to make lib-support.h and loader.c more readable.
And also have many ways to refactor these messy code.
BTW, aarch64 is ready for coding, but I may have no more time to do this.

**There are also some ARM relocation issue**
1. REL instead of RELA
2. Global variables appear in .rel.dyn force our code needs to add WRITE permission in .text section
    >> maybe some compiler option can avoid, I'm not sure

### CODE REVIEW

File: loader.c [hunk 1]
```diff
@@ -293,13 +306,35 @@ dloader_p api_load(const char *filename)
     ElfW_Reloc *relocs =
         (ElfW_Reloc *)(load_bias + get_dynamic_entry(dynamic, ELFW_DT_RELW));
     size_t relocs_size = get_dynamic_entry(dynamic, ELFW_DT_RELWSZ);
+
+    /*
+     * FIXME
+     * There is no RELA in ARM instead of REL,
+     * someone should make the condition code better.
+     */
+    if(relocs_size == 0) {
```
Comment (jserv) [inline loader.c]: Follow consistent coding style.

**Review by jserv — COMMENTED** [review]
Why would you do this? Can't you specify build target in `Makefile`?

File: lib-support.h [hunk 2]
```diff
@@ -3,6 +3,81 @@
...
+#define REG_ARG_2   {r1}
+#define REG_RET  r0
+#define LABEL_PREFIX  "="
+#define SYS_ADDR_ATTR "word"
+
+#define PUSH_STACK_STATE "push {r11, lr}"
+#define POP_STACK_STATE  "pop {r11, lr}"
+
+#elif defined(__aarch64__)
+#define _PUSH(x,y) stp x, y, [sp, #-16]!
```
Comment (jserv) [inline lib-support.h]: aarch64 support is incomplete, right?

**Comment by lecopzer** [comment]
Sorry I didn't create an issue first.

Why I did this:
1. There is a lot of  `#if defined(__X86_64__)` in source code. It no need to have this macro in every arch-specific code if you want to support only X86_64. 
You can block unsupported arch in Makefile to clean the code.

2. Why not support ARM? The code builds well to support with new architecture.
In best result after porting new arch, `loader.c` can have no extra code after resolve `FIXME`, and system-specific macro in lib-support.h can separate into something like `arch/arm.h  arch/x86_64.h`.
The original source code after porting will not be too much different.

In summary, make `min-dl ` more portability.

/////
I can support ARM in Makefile like the code in follows
```
CROSS_COMPILE_SUFFIX ?= -linux-gnueabi-
arm: 
  CROSS_COMPILE = $@$(CROSS_COMPILE_SUFFIX)
  make all
```
I can fix what you mentions after getting off work

File: Makefile [hunk 3]
```diff
@@ -1,9 +1,10 @@
-CFLAGS = -std=gnu99 -Wall -Werror -g -D_GNU_SOURCE
+CC      = $(CROSS_COMPILE)gcc
+CFLAGS  = -std=gnu99 -Wall -Werror -g -D_GNU_SOURCE
 CFLAGS += -DPROG_HEADER=prog_header
 
-OUT = out
+OUT     = out
```
Comment (jserv) [inline Makefile]: Don't change this line. Minimize the area of modifications.

File: loader.c [hunk 4]
```diff
@@ -117,7 +118,17 @@ static int prot_from_phdr(const ElfW(Phdr) *phdr)
         prot |= PROT_WRITE;
     if (phdr->p_flags & PF_X)
         prot |= PROT_EXEC;
+    /*
+     * FIXME
+     * In ARM, some global variables in .text will appear in .rel.dyn,
+     * and we need to give it WRITE permission to relocate it.
+     * It must be fixed due to security issue !!!
```
Comment (jserv) [inline loader.c:117]: Would you like to fix as well?

File: loader.c [hunk 5]
```diff
@@ -198,11 +207,9 @@ dloader_p api_load(const char *filename)
         fail(filename, "File has no valid ELF header!", NULL, 0);
 
     switch (ehdr.e_machine) {
-#if defined(__x86_64__)
     case EM_X86_64:
-#else
-#error "Don't know the e_machine value for this architecture!"
-#endif
+    case EM_ARM:
+    case EM_AARCH64:
```
Comment (jserv) [inline loader.c]: If Aarch64 support is incomplete, don't enumerate.

**Comment by jserv** [comment]
Use `git rebase -i` to squash comments in order to get rid of incomplete Aarch64 code.

File: loader.c [hunk 6]
```diff
@@ -273,8 +284,8 @@ dloader_p api_load(const char *filename)
                          start - last_page_end, PROT_NONE);
 
             mmap((void *) start, end - start,
-                 prot_from_phdr(ph), MAP_PRIVATE | MAP_FIXED, fd,
-                 round_down(ph->p_offset, pagesize));
+              prot_from_phdr(ph), MAP_PRIVATE | MAP_FIXED, fd,
```
Comment (jserv) [inline loader.c]: Avoid such syntax changes.

File: loader.c [hunk 7]
```diff
@@ -293,13 +304,29 @@ dloader_p api_load(const char *filename)
     ElfW_Reloc *relocs =
         (ElfW_Reloc *)(load_bias + get_dynamic_entry(dynamic, ELFW_DT_RELW));
     size_t relocs_size = get_dynamic_entry(dynamic, ELFW_DT_RELWSZ);
+
+    /*
+     * FIXME
+     * There is no RELA in ARM instead of REL,
+     * someone should make the condition code better.
```
Comment (jserv) [inline loader.c]: Skip the line starting with "someone". `FIXME` section only denotes further modifications.

File: loader.c (hunk 4 above)
Comment (lecopzer) [reply loader.c:117]: OK I'll try it.
but it better to open an new issue

**Comment by lecopzer** [comment]
OK, I fix the issue you mentioned except global variable issue
Please check it.
THX

File: loader.c [hunk 8]
```diff
@@ -293,13 +297,18 @@ dloader_p api_load(const char *filename)
...
         switch (reloc_type) {
-#if defined(__x86_64__)
         case R_X86_64_RELATIVE:
-#endif
+        {
+            ElfW(Addr) *addr = (ElfW(Addr) *)(load_bias + reloc->r_offset);
+            *addr += load_bias;
+            break;
+        }
+        case R_ARM_RELATIVE:
```
Comment (jserv) [inline loader.c]: Is it ARM specific?

Comment (lecopzer) [reply loader.c]: Yes, should I revise as
 ```
case R_X86_64_RELATIVE:
case R_ARM_RELATIVE:
         {
             ElfW(Addr) *addr = (ElfW(Addr) *)(load_bias + reloc->r_offset);
             *addr += load_bias;
             break;
         }
```

Comment (lecopzer) [reply loader.c]: e22bf56

**Comment by jserv** [comment]
Follow the instructions to use `git rebase -i`: https://blog.yorkxin.org/2011/07/29/git-rebase

For example, commit `ARM: Support for ARM and x86_64` and commit `loader: delete all debug and unfinished arch` should be squashed into single commit. And, commits for several syntax tweaks should not exist.

**Comment by lecopzer** [comment]
I just rebase `ARM: Support for ARM and x86_64` ,  `loader: delete all debug and unfinished arch` and `loader: Fixed minor coding style...`
And also Makefle --- Minimize the area of modifications.
Is this what you mean?

loader.c is 353 line to 359 line, looks good

**Comment by jserv** [comment]
sort of. Similarly, commit `lib-support: Make arm assembly works in C Macro` and `lib-support: Fixed wrong macro expanded` should be squashed as well.

**Review by jserv — COMMENTED** [review]
The subject can be changed to "Support ARM architecture", which is simpler.

File: loader.c [hunk 9]
```diff
@@ -165,17 +168,15 @@ ElfW(Word) get_dynamic_entry(ElfW(Dyn) *dynamic, int field)
...
-    "plt_trampoline:\n"
-    "pop %rdi\n" /* Argument 1 */
-    "pop %rsi\n" /* Argument 2 */
-    "call system_plt_resolver\n"
-    "jmp *%rax\n"
-    ".popsection\n");
-#else
-#error "Unsupported architecture"
-#endif
+asm(".pushsection .text,\"ax\",\"progbits\""  "\n"
```
Comment (jserv) [inline loader.c:175]: More comments are necessary for maintenance purpose since the crucial parts are now moved to `arch/` directory along with extra macros.

File: lib-support.h [hunk 10]
```diff
@@ -3,6 +3,36 @@
...
+
+#if defined(__x86_64__)
+#include "arch/x86_64.h"
+#elif defined(__arm__)
+#include "arch/arm.h"
+#else
+#error "Unsupported architecture"
+#endif
+
+#define PUSH_S(x)        XSTR(_PUSH_S(x))
```
Comment (jserv) [inline lib-support.h]: Ditto. Comments are necessary for maintenance purpose.

**Comment by lecopzer** [comment]
I add the comment in 7930411
Is the comment okay for maintenance purpose?

**Comment by lecopzer** [comment]
This way?

**Review by jserv — COMMENTED** [review]
Change the subject to "Support ARM architecture", and the body of commit messages is meaningless while your mentioning on `FIXME`.

**Review by jserv — COMMENTED** [review]
You can change the subject of Git commit messages to "Split architecture-specific parts into separate directories".

**Review by jserv — COMMENTED** [review]
The subject is misleading. Instead, it is nothing other than modularizing.

**Comment by lecopzer** [comment]
Okay I am done.
It seems that I have no improvement in these years....So sad

**Comment by jserv** [comment]
Thank @lecopzer again for this contribution!
